scripts/clean_osm_data.py

```diff
@@ -279,279 +288,320 @@ def finalize_lines_type(df_lines):
+        if not row[col_to_fill]:
+            return None
+        n_missing = size_to_fill - len(row[col_to_fill])
+        fill_val = row[col_to_fill][-1]
 
-    # Keep original frequency if it exists  and set a standard value
-    # NB The standard frequency value may be regional-dependent
-    # TODO Fill by adjancent value
+        new_row = row[col_to_fill].copy()
 
-    # Initialize a default frequency value
-    ac_freq_default = 50.0
+        return row[col_to_fill] + [fill_val] * n_missing
 
-    if "tag_frequency" in df_all_lines.columns:
-        ac_grid_freq_levels = (
-            df_all_lines["tag_frequency"]
-            .value_counts(sort=True, dropna=True)
-            .drop(["0", 0], errors="ignore")
-        )
+    df_fhv = df[len_freq > len_voltage].apply(
+        lambda row: _fill_by_last(row, "voltage", "tag_frequency"),
+        axis=1,
+    )
+    df.loc[df_fhv.index, "voltage"] = df_fhv
 
-        if not ac_grid_freq_levels.empty:
-            ac_freq_default = float(ac_grid_freq_levels.idxmax())
+    df_vhf = df[len_freq < len_voltage].apply(
+        lambda row: _fill_by_last(row, "tag_frequency", "voltage"),
+        axis=1,
+    )
+    df.loc[df_vhf.index, "tag_frequency"] = df_vhf
 
-        df_all_lines.tag_frequency.fillna(ac_freq_default, inplace=True)
+    return df
 
-        df_all_lines["tag_frequency"] = (
-            df_all_lines.tag_frequency.astype(str)
-            .str.split(pat=";", n=1)
-            .str[0]
-            .astype(float)
-        )
 
-    # Add frequency column if not present in data
-    else:
-        df_all_lines["tag_frequency"] = ac_freq_default
-
-    df_all_lines = split_cells_multiple(df_all_lines)
-    # Add circuits information
-    # if not int make int
-    if df_all_lines["cables"].dtype != int:
-        dropped_cables = [
-            x for x in dropped_cables_tags if x in df_all_lines["cables"].values
-        ]
-        if len(dropped_cables) != 0:
-            logger.info(
-                f"The lines with a cables tag in {set(dropped_cables)} will be dropped."
-            )
-
-        # map known non-numerical issues into a reasonable n_cables value
-        df_all_lines["cables"] = (
-            df_all_lines["cables"]
-            .map(cables_tag_to_n_cables)
-            .fillna(df_all_lines["cables"])
-        )
-        # HERE. "0" if cables "None", "nan" or "1"
-        df_all_lines.loc[
-            (df_all_lines["cables"] < "3") | df_all_lines["cables"].isna(), "cables"
-        ] = "0"
-
-        # there may be some non-known numerical issues
-        not_resolved_cables = (
-            ~pd.to_numeric(df_all_lines["cables"], errors="coerce")
-            .astype(float)
-            .apply(float.is_integer)
-        )
-        unknown_cables_tags = set(
-            df_all_lines.loc[not_resolved_cables]["cables"].values
+def fill_circuits(df):
+    """
+    This function fills the circuits column to match the length of v/f columns
```

Comment:
This function fills the circuits column to match the length of v/f columns OR
This function fills the rows in the circuits column to match the length of the rows in the voltage and frequency columns?

Reply:
exactly.
I added a large docstring